zokrates_core/src/static_analysis/zir_propagation.rs

```diff
@@ -0,0 +1,516 @@
+use crate::zir::{
+    BooleanExpression, FieldElementExpression, Folder, UExpression, UExpressionInner, Variable,
+    ZirExpression, ZirFunction, ZirProgram, ZirStatement,
+};
+use std::collections::HashMap;
+use zokrates_field::Field;
+
```

Comment:
Is there a reason we couldn't use the zir Folder here? Then we just need to implement folding on `ZirStatement::Definition` (adding to the constant map) and `ZirExpression` (reading from the constant map)

Reply:
The rest would come from the default implementation of the folder